Start conversion to easy-smt

sampsyo opened the pull request:
As discussed on Slack, here's a first attempt to move from rsmt2 to [easy-smt](https://github.com/elliottt/easy-smt). Here are the high-level takeaways:

* The actual interaction with the solver was easy to move over. No problem! Just had to fiddle around with the APIs to line them up.
* The real work here is the drudgery of changing from plain ol' strings for the SMT expressions to the library's `SExpr` type.

I got everything in `solver.rs` converted, including all the SMT generation and the interaction with the solver. This took me about 3 hours to move everything over—almost all of that was replacing calls to `format!` with explicit S-expression construction calls. It's not exactly fun, but it's not exactly _hard_ either.

There is more drudgery ahead: the files in `encoded_ops` have a lot of raw SMT strings in them and will be a lot of work to convert.

There is a chance I have made a terrible mistake and I should not have tried to do all this annoying conversion at all: instead, we could try to _parse_ the strings we're _already_ generating and turn them into `SExpr`s. Or we could attempt to hack `easy-smt` to allow us to shove strings directly into the solver, which it doesn't currently allow (you have to provide an `SExpr`). Both of those seem very reasonable in retrospect, but now I wonder if either of them would be simpler than just forging ahead and finishing the conversion. I'd be interested in others' opinions, given what the code looks like in `solver.rs` after the conversion.

sampsyo committed:
Extremely crude easy-smt replacement

I tried to do this as fast as I could, not trying to clean anything up
after the transition, and just get the replacement done as quickly as
possible. (And it still took me about 3 hours.) The main work, of
course, is replacing all the raw string manipulation (mainly `format!`
calls) with S-expression builder calls.

sampsyo committed:
Try attempting to convert clz encoding...

There is a lot of typing to be done here!

sampsyo committed:
Try pre-declaring some atoms

sampsyo committed:
Add a `declare` helper

avanhatt commented on cranelift/isle/veri/veri_engine/src/solver.rs:
```diff
@@ -867,12 +938,12 @@ pub fn run_solver(rule_sem: RuleSemantics, query_width: usize) -> VerificationRe
     }
 
-    let assumption_str = format!("(and {})", assumptions.join(" "));
+    let assumption_conjunction = solver.and_many(assumptions);
```
nice

sampsyo committed:
Temporarily disable encoded ops

I'm trying to get to a place where the crate compiles, even though the
encoded ops don't work at all.

sampsyo committed:
Satisfy the borrow checker

sampsyo commented:
As we discussed synchronously today, I've mucked things up so that our crate *builds*. That meant temporarily:

* removing the files under `encoded_ops`
* replacing the calls into their SMT-generating functions with `todo!()`

It works! I haven't tried running any tests, because I don't feel qualified to know which tests should be runnable in this semi-broken state. (I could probably figure that out, but I'm going to move on to trying to rewrite the encoded ops stuff instead of spending time on that.)

sampsyo committed:
Try converting one rev function

sampsyo committed:
Add a simple `assume` helper

sampsyo committed:
Update to latest easy-smt

Now using the builder to construct contexts. And there's a cool new
`numeral` helper for building number literals of all kinds!

sampsyo committed:
Add back `rev` with one working function & test

sampsyo committed:
More reusable test framework

avanhatt commented on cranelift/isle/veri/veri_engine/src/solver.rs:
```diff
@@ -250,44 +320,45 @@ impl SolverCtx {
     }
 
-    pub fn vir_to_rsmt2_constant_ty(&self, ty: &Type) -> String {
+    pub fn vir_to_rsmt2_constant_ty(&self, ty: &Type) -> SExpr {
```
nit: remove `rsmt2` from name

sampsyo replied:
Thanks; addressed!

sampsyo committed:
Add encoding files from @mpardesh

sampsyo committed:
Minor Python cleanup

sampsyo committed:
Generate easy-smt code for declarations

sampsyo committed:
Convert assertions to easy-smt

sampsyo committed:
Slight Python cleanup

sampsyo committed:
Bind Rust variables for SMT variable names (atoms)

sampsyo committed:
Slightly simplify some generated Rust

sampsyo committed:
Finish test for rev1

sampsyo committed:
Fix variable names

Apparently, `findall` returns a list of strings (not tuples) when there
is only one subgroup.

sampsyo committed:
Use generated code for rev1

sampsyo committed:
Generate code for remaining rev*

sampsyo committed:
Re-enable calls to `rev*`

sampsyo committed:
Rename cls* SMT files for consistency

To match clz*.

sampsyo committed:
Clean up test SMT declarations

sampsyo committed:
Use proper convention for return variables

sampsyo committed:
Split out `a64cls32` from `cls32` SMT file

Quoth @mpardesh on Slack:

> also note that the cl32 file encodes cls for 32 bits and the a64 version
> after a bunch of empty lines

Now it's two separate files.

sampsyo committed:
Don't preserve blank lines anymore

Makes it a little easier to copy n' paste into the Rust file.

sampsyo committed:
Start adding back cls.rs

sampsyo committed:
Add a `cls1` file for uniformity

I know this is not a very complicated blob of SMT, but now we have a 1:1
ratio of Rust functions to SMT input files.

sampsyo committed:
Convert cls1

sampsyo committed:
Generate cls8

sampsyo committed:
Generate cls8 code

sampsyo committed:
Finish generating cls16

sampsyo committed:
Generate cls32

sampsyo committed:
Finish generating cls

sampsyo committed:
Fix a bug in cls1

sampsyo committed:
Remove needless import

sampsyo committed:
Fix a variable reference in a64cs32

sampsyo committed:
Enable calls to cls* functions

sampsyo committed:
Fix return variable names for clz*

sampsyo committed:
Add clz1 SMT input

sampsyo committed:
Restore clz.rs and generate clz1

sampsyo committed:
Generate clz.rs, hopefully

But not a64clz32... don't have a file for this one?

sampsyo committed:
Rearrange some `todo!()` calls

sampsyo committed:
Remove an unnecessary `mut`

sampsyo committed:
Add extensions to the SMT files

sampsyo committed:
Add a README for encodings generation

sampsyo commented:
In the name of expediency—that is, to get something actually working as quickly as possible, so we can iterate on it—I had to change tactics a bit. The original plan had been to port the generation code from Python to Rust, so we could generate S-expressions directly (instead of generating Rust that generates S-expressions). But what I didn't quite realize is that the "source of truth" is in SMT-LIBv2 code already, and that the Python code acts as a translator for *that* code.

In light of that, I saw two fast routes to getting something working:

1. Still port the Python to Rust, but that Rust code would work by parsing the SMT-LIBv2 files (just as the current Python does).
2. Keep the status quo (Python generates Rust that generates S-expressions), but change the generator to produce Rust code that produces `easy-smt` `SExpr` values instead of strings.

I opted for option 2 because getting an S-expression parser going on the Rust side seemed like a big drag. We may want to revisit this later.

Logistically, my goal was to keep the "source of truth" in the original templatized SMT-LIBv2 files—not in the generated Rust code. With that philosophy, if we need to fix a bug in the encoding, we never fix the generated Rust; we fix the original SMT S-expressions (or the Python converter) and then just regenerate the Rust. The goal, then, is to have no (or at least minimal) human massaging after conversion. To that end, I did these things:

* Checked in the input SMT files and the Python conversion script into git (see the `encodings` directory).
* Hacked the Python script to produce `easy-smt`-based generator code instead of the string-based stuff.
* Organized the SMT files so we have a 1:1 mapping between those files and the Rust functions that wrap them. That means that I split up `a64cls32` from `cls32` so it has its own file, and I added `cl*1` files (even though they are super simple and you could just write them directly in Rust instead).
* I tried to delimit the generated code the Rust files in such a way that it is hopefully easy to update. You just need to do `python3 convert.py cls16.smt2 | pbcopy`, for example, and paste this into the corresponding Rust function body.

Here are some loose ends to address next:

* I wasn't quite able to grok where the SMT-LIBv2 code lives for two oddball functions: `rbit32` and `a64clz32`. Perhaps @mpardeshi has these sitting around?
* We should probably add annotations to quell Clippy for generated code (it's not meant for human consumption anyway)?
* There were several cases where the current Rust code didn't seem to correspond exactly to the original SMT-LIBv2 sources I was working from, so I had to make some guesses. It is extremely likely that I broke something when doing that, so we need to check whether any of these functions still actually work.

One thing I'm not sure about is: What is the best way to check whether these encodings are working at all? Even if it's not an automatable test, being able to check whether these encodings work in isolation seems really handy… especially if it can happen separately from the rest of the verification machinery.

Less-urgent things that would be good to do but not in this PR include:

* The generated `SExpr`-construction code could be improved. It doesn't use as many helpers as it could, instead opting to use mostly the low-level utilities.
* We could write some tests. (I started writing one test for `rev1`, the simplest possible thing, and then abandoned it.)
* We could make things much more convenient without rocking the boat by generating entire Rust files from scratch (including function definitions and extract/concat calls) so no humans are involved at all to copy & paste.
* We could consider ripping this Python approach out entirely and replacing it with direct parsing of the SMT-LIBv2 sources in Rust.
* Or we could, at some point, instead shift the "source of truth" to the generated Rust. We would throw away the SMT-LIBv2 sources and consider the Rust code to be canonical. Maybe this makes it way more annoying to tweak; I'm not sure.

sampsyo committed:
Remove dangling references to rsmt2

mpardesh commented:
Woohoo that was super fast!!!

Here are some quick, minor comments:
  - I think `rev32` should be in the Slack thread
  - Ah yes, I think I couldn't find a64clz32 but I'll look more and add it to the thread
  - Sorry, should've mentioned I didn't make files for the 1 bit functions because they were just a line or two of manual Rust
  - To test: you can check for equivalence of the raw smt and C test file (with some input size changing in C). Beyond that just used the regular tests in `veri_engine/examples`

sampsyo committed:
Add a64clz32

sampsyo committed:
Generate a64clz32

sampsyo commented:
Awesome; thanks! As for testing specifically, I suppose I want to try setting up something (semi-)automated, since I am not very good at getting the details right when manually setting things up for one-off tests… still not 100% sure what that would look like. Maybe running the solver from an isolated Rust test?

avanhatt commented:
Something we could try for testing (that would be good for testing the rules in general):

1. Set up a solver instance in the test.
2. Set up SMT variables and set them equal to specific inputs. Check that the output of the function from the model is what we expect. For more assurance, we could also do a CEGIS-like thing and check that there is no model such that the output is _not_ what we expect.
3. We could set up nice helpers for the above, so that it can be called over multiple inputs. 

But, I don't think this needs to happen in this PR.

avanhatt committed:
Merge branch 'verify-main' into easy-smt

avanhatt commented:
Other important TODO before this becomes the default: `easy-smt`'s debugging doesn't list the actual values. We should make sure things like this get displayed in a more useful way:
```
Assertion list is feasible
LHS and RHS equality condition:
        SExpr::List(14853)            // <- this should have named values
```

avanhatt commented:
^ Doing a pass to add `ctx.smt.display(...)` calls.

avanhatt committed:
better printing/logging

avanhatt commented:
Currently failing tests:

failures:
[2363](https://github.com/avanhatt/wasmtime/actions/runs/4007164981/jobs/6879644553#step:7:2364)
    test_broken_uextend
[2364](https://github.com/avanhatt/wasmtime/actions/runs/4007164981/jobs/6879644553#step:7:2365)
    test_clz
[2365](https://github.com/avanhatt/wasmtime/actions/runs/4007164981/jobs/6879644553#step:7:2366)
    test_ctz
[2366](https://github.com/avanhatt/wasmtime/actions/runs/4007164981/jobs/6879644553#step:7:2367)
    test_ctz_broken
[2367](https://github.com/avanhatt/wasmtime/actions/runs/4007164981/jobs/6879644553#step:7:2368)
    test_sextend
[2368](https://github.com/avanhatt/wasmtime/actions/runs/4007164981/jobs/6879644553#step:7:2369)
    test_uextend

avanhatt commented on cranelift/isle/veri/veri_engine/src/solver.rs:
```diff
@@ -336,35 +405,35 @@ impl SolverCtx {
                     }
-                    Type::Int => i.to_string(),
+                    Type::Int => self.smt.numeral(i),
                     Type::Bool => {
```
Aha, I think this is causing the failing tests! Will fix.

sampsyo replied:
Oh Lordy, that is obviously wrong!! Nice catch!

avanhatt committed:
fix boolean literals, better model display

avanhatt committed:
Add rev tests that hit the solver

sampsyo committed:
Try writing rbit32 in raw SMT

Based on @mpardeshi's advice, this is just a copy of `rev32` but with an
extra extract at the top. I hope this works.

sampsyo committed:
Generate rbit32

That's the last one, hopefully!

avanhatt committed:
Run tests with timeout and retry attempt

avanhatt committed:
actually retry, longer timeout

avanhatt committed:
Merge branch 'verify-main' into easy-smt

avanhatt committed:
fix test output on failed test

avanhatt commented:
Ok, I think the `clz` and `ctz` encodings might actually be broken. Investigating!

avanhatt committed:
Add more encoding tests (some failing)

avanhatt committed:
more

mpardesh commented on cranelift/isle/veri/encodings/clz16.smt2:
```diff
@@ -0,0 +1,39 @@
+
+(assert (= y1_{id} (bvlshr x2_{id} #x0001)))
+(assert (ite (not (= y1_{id} (_ bv0 16))) (= ret5_{id} ret4_{id}) (= ret5_{id} (bvadd ret4_{id} (_ bv1 16)))))
+(assert (ite (not (= y1_{id} (_ bv0 16))) (= x1_{id} y1_{id}) (= x1_{id} x2_{id})))
```
after this line, try adding 
```
; last round
(declare-fun ret6_{id} () (_ BitVec 16))
(assert (ite (not (= x1_{id} (_ bv0 16))) (= ret6_{id} ret5_{id}) (= ret6_{id} (bvadd ret5_{id} (_ bv1 16)))))
```
and then regenerating?

avanhatt replied:
Aha, good catch! Trying this now

mpardesh commented on cranelift/isle/veri/encodings/clz32.smt2:
```diff
@@ -0,0 +1,48 @@
+
+(assert (= y1_{id} (bvlshr x2_{id} #x00000001)))
+(assert (ite (not (= y1_{id} (_ bv0 32))) (= ret5_{id} ret4_{id}) (= ret5_{id} (bvadd ret4_{id} (_ bv1 32)))))
+(assert (ite (not (= y1_{id} (_ bv0 32))) (= x1_{id} y1_{id}) (= x1_{id} x2_{id})))
```
after this line try adding:
```
; last round
(declare-fun ret6_{id} () (_ BitVec 32))
(assert (ite (not (= x1_{id} (_ bv0 32))) (= ret6_{id} ret5_{id}) (= ret6_{id} (bvadd ret5_{id} (_ bv1 32)))))
```
and then regenerating?

mpardesh commented on cranelift/isle/veri/encodings/clz64.smt2:
```diff
@@ -0,0 +1,57 @@
+
+(assert (= y1_{id} (bvlshr x2_{id} #x0000000000000001)))
+(assert (ite (not (= y1_{id} (_ bv0 64))) (= ret6_{id} ret5_{id}) (= ret6_{id} (bvadd ret5_{id} (_ bv1 64)))))
+(assert (ite (not (= y1_{id} (_ bv0 64))) (= x1_{id} y1_{id}) (= x1_{id} x2_{id})))
```
after this line try adding
```
; last round
(declare-fun ret7_{id} () (_ BitVec 64))
(assert (ite (not (= x1_{id} (_ bv0 64))) (= ret7_{id} ret6_{id}) (= ret7_{id} (bvadd ret6_{id} (_ bv1 64)))))
```
and then regenerating?

avanhatt committed:
final rounds for clzs

avanhatt committed:
Merge branch 'verify-main' into easy-smt

avanhatt committed:
fix 64 case, add tests for cls encoding
